### PR TITLE
Upgrade to java11

### PR DESCRIPTION
**Pull request by lindseydew**
## What does this change?

This upgrades the lambdas that serve the Save For Later functionality from Java8 to Java11

## How to test
Tested Save for Later on CODE:

- [x] Can get articles
- [x] Can post articles

Tested User Deletion CODE:

- [x] Delete a user account and see that this is deleted from Save For Later

Note I wasn't able to test the full interaction on CODE as I have been unable to find the exact string the app is expecting to parse. I think the json parsing logic needs a bit of an update too, so I'll make a note to do that and this should make it possible to test an example event. 

For now, we have alarms set up on PROD if the delete interaction is unable to complete. 

## Accessibility

-   [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-   [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-   [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-   [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

### CODE REVIEW

File: project/dependencies.scala [hunk 1]
```diff
@@ -7,28 +7,35 @@ object Dependencies {
...
-  val jackson = "com.fasterxml.jackson.module" %% "jackson-module-scala" % jacksonVersion
-  val jacksonDataFormat = "com.fasterxml.jackson.dataformat" % "jackson-dataformat-cbor" % jacksonVersion
-  val jacksonJdk8DataType = "com.fasterxml.jackson.datatype" % "jackson-datatype-jdk8" % jacksonVersion
-  val jacksonJsrDataType = "com.fasterxml.jackson.datatype" % "jackson-datatype-jsr310" % jacksonVersion
+  val jackson =
+    "com.fasterxml.jackson.module" %% "jackson-module-scala" % jacksonVersion
+  val jacksonDataFormat =
+    "com.fasterxml.jackson.dataformat" % "jackson-dataformat-cbor" % jacksonVersion
+  val jacksonJdk8DataType =
+    "com.fasterxml.jackson.datatype" % "jackson-datatype-jdk8" % jacksonVersion
```
Comment (mchv) [inline project/dependencies.scala:21]: I think this may be removed

**Comment by mchv** [comment]
@lindseydew I noticed a few additional changes linked to this upgrade which I have directly added as additional commits..but the build is now breaking 😢! I think this is because `sbt-assembly` had changed its API, and it should not be too hard to [fix existing file](https://github.com/guardian/mobile-save-for-later/blob/main/project/MergeFilesStrategy.scala) [using documentation](https://github.com/sbt/sbt-assembly?tab=readme-ov-file#merge-strategy) (but I have not done it!).

Anyway once this is fixed, the upgrade to `java 21` should be straightforward as it supported by latest scala `2.12.x` and sbt `1.9.9`, and so require only a few changes.

**Comment by lindseydew** [comment]
> @lindseydew I noticed a few additional changes linked to this upgrade which I have directly added as additional commits..but the build is now breaking 😢! I think this is because `sbt-assembly` had changed its API, and it should not be too hard to [fix existing file](https://github.com/guardian/mobile-save-for-later/blob/main/project/MergeFilesStrategy.scala) [using documentation](https://github.com/sbt/sbt-assembly?tab=readme-ov-file#merge-strategy) (but I have not done it!).
> 
> Anyway once this is fixed, the upgrade to `java 21` should be straightforward as it supported by latest scala `2.12.x` and sbt `1.9.9`, and so require only a few changes.

Great thank you, I'll take a look at the build

File: mobile-save-for-later-user-deletion/template.yml [hunk 2]
```diff
@@ -0,0 +1,84 @@
+# This AWS SAM template has been generated from your function's configuration. If
```
Comment (waisingyiu) [inline mobile-save-for-later-user-deletion/template.yml:1]: Is it the cloudformation for mobile-save-for-later?  Why do we need this one when there is "conf/cfn.yaml"?

File: mobile-save-for-later/template.yml [hunk 3]
```diff
@@ -0,0 +1,97 @@
+# This AWS SAM template has been generated from your function's configuration. If
```
Comment (waisingyiu) [inline mobile-save-for-later/template.yml:1]: I think CDK is used to define the cloudformation stack for this service?  Is it used for another purpose?

File: build.sbt [hunk 4]
```diff
@@ -49,10 +49,48 @@ val commonSettings: immutable.Seq[Def.Setting[_]] = List(
     specsScalaCheck,
     specsMock
   ),
-  assembly / assemblyMergeStrategy := {
+  ThisBuild / assemblyMergeStrategy := {
     case "META-INF/MANIFEST.MF" => MergeStrategy.discard
-    case "META-INF/org/apache/logging/log4j/core/config/plugins/Log4j2Plugins.dat" =>
-      new MergeFilesStrategy
+    case PathList(ps @ _*) if ps.last equalsIgnoreCase "Log4j2Plugins.dat" =>
```
Comment (waisingyiu) [inline build.sbt:54]: You may have explained this in our snyk session, but I am not sure why we change the implementation of merge strategy here?  The previous one does not work?

Comment (lindseydew) [reply build.sbt:54]: yeah, I was getting an error with instantiating the MergeFilesStrategy class.

File: mobile-save-for-later/template.yml (hunk 3 above)
Comment (lindseydew) [reply mobile-save-for-later/template.yml:1]: Yeah, we are using the CDK for the cfn stack definition. This is used to be able to run the lambda functions locally using aws sam. Instructions here: https://github.com/guardian/mobile-save-for-later/pull/102/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R86

**Review by waisingyiu — APPROVED** [review]
Thank you for the great work, I am happy to approve.

As an aside, I suppose that I need a docker desktop license if I want to run this service locally?  (I see your name on the docker desktop license spreadsheet)

File: build.sbt [hunk 5]
```diff
@@ -63,12 +101,11 @@ val commonSettings: immutable.Seq[Def.Setting[_]] = List(
...
   organization := "com.gu",
   version := "1.0",
-  scalaVersion := "2.12.5",
+  scalaVersion := "2.12.18",
   scalacOptions ++= Seq(
     "-deprecation",
     "-encoding",
     "UTF-8",
-    "-target:jvm-1.8",
     "-Ypartial-unification",
```
Comment (rtyley) [inline build.sbt:109]: It's not essential, but I quite like adding `-release:11` to the `scalacOptions`, as a way of sbt ensuring that it's being run with at least Java 11 or above:

```suggestion
    "-release:11",
    "-Ypartial-unification",
```

...the `-release:11` flag has the effect of ensuring that the generated class files are suitable for being run on a Java 11 runtime.

**Review by rtyley — COMMENTED** [review]
I've only quickly skimmed through the diff of this PR, as I'm a bit short of time today! 

I think the diff might be a bit bigger than it needs to be, I think because `scalafmt` formatting changes are being included - that makes it a bit harder to concentrate on the significant code changes? If scalafmt is desired, it might be worth doing a separate scalafmt PR and then rebasing this one on top of that.

@lindseydew, you suggested that maybe should merge this PR, and then merge https://github.com/guardian/mobile-save-for-later/pull/105, which is understandable because https://github.com/guardian/mobile-save-for-later/pull/105 talks about Java 21, and this PR is for Java 11!

However, https://github.com/guardian/mobile-save-for-later/pull/105 only _allows_ `mobile-save-for-later` to build under Java 21, doesn't _require_ it, and doesn't even attempt to update `mobile-save-for-later` up from Java 8 - consequently, it's a much smaller PR, easier to review, and therefore, I would say, much lower risk!

So personally, I would prefer to merge https://github.com/guardian/mobile-save-for-later/pull/105 immediately, and then look at rebasing this PR on top of main - maybe even separating out a scalafmt run into it's own PR first. I'd be really happy to pair tomorrow on those stages of the operation if that would help?

**Comment by lindseydew** [comment]
> I've only quickly skimmed through the diff of this PR, as I'm a bit short of time today!
> 
> I think the diff might be a bit bigger than it needs to be, I think because `scalafmt` formatting changes are being included - that makes it a bit harder to concentrate on the significant code changes? If scalafmt is desired, it might be worth doing a separate scalafmt PR and then rebasing this one on top of that.
> 
> @lindseydew, you suggested that maybe should merge this PR, and then merge #105, which is understandable because #105 talks about Java 21, and this PR is for Java 11!
> 
> However, #105 only _allows_ `mobile-save-for-later` to build under Java 21, doesn't _require_ it, and doesn't even attempt to update `mobile-save-for-later` up from Java 8 - consequently, it's a much smaller PR, easier to review, and therefore, I would say, much lower risk!
> 
> So personally, I would prefer to merge #105 immediately, and then look at rebasing this PR on top of main - maybe even separating out a scalafmt run into it's own PR first. I'd be really happy to pair tomorrow on those stages of the operation if that would help?

Ok sounds good. Yes I'm happy for you to merge your branch first. Agree with separating the PR. A pairing session would be great. When would you be free?

**Comment by lindseydew** [comment]
Closing because this has been superseded by subsequent PRs: https://github.com/guardian/mobile-save-for-later/pull/106 and https://github.com/guardian/mobile-save-for-later/pull/107

**Pull request closed without merging**